PHOENIX-6960 Scan range is incorrect when query desc columns

virajjasani opened the pull request:
Jira: PHOENIX-6960

virajjasani commented:
build results: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1663/4/testReport/

stoty commented on phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:
```diff
@@ -1722,7 +1725,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
+            boolean lowerInclusive = column.getSortOrder() == SortOrder.ASC;
+            boolean upperInclusive = column.getSortOrder() == SortOrder.DESC;
+            KeyRange range = type.getKeyRange(lowerRange, lowerInclusive, upperRange,
+                    upperInclusive, SortOrder.ASC);
```
I have debugged into this, and after inverting twice, we end up returning a non-inverted range, which eventually gets inverted in WhereOptimizer.pushKeyExpressionsToScan().

Do we even need to take SortOrder into account here, and if yes, couldn't we simplify this logic ?

stoty replied:
I'm talking the whole method, not the specific line above.

stoty commented on phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:
```diff
@@ -1691,8 +1691,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
+            // TODO: is there a case where we'd need to go through the childPart to calculate the key range?
+            PColumn column = childSlot.getKeyPart().getColumn();
+            PDataType type = column.getDataType();
+            byte[] key = PVarchar.INSTANCE.toBytes(startsWith, column.getSortOrder());
```
This is where we get an inverted range

stoty commented on phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:
```diff
@@ -1722,7 +1725,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
+            KeyRange range = type.getKeyRange(lowerRange, lowerInclusive, upperRange,
+                    upperInclusive, SortOrder.ASC);
             if (column.getSortOrder() == SortOrder.DESC) {
                 range = range.invert();
```
And this is wehere we re-invert, getting a normal range back.

virajjasani commented:
Thank you for taking a look @stoty, and yes you are correct that we have scope for optimization here, i just addressed your review with the latest revision. Also, @jinggou found some interesting test case failures, i tried to address them.

Created 5.1 backport PR: #1668

@stoty @jinggou could you please take a look again? (not urgent, i might likely be able to come online after 5 days)

stoty reviewed (approved):
+1 LGTM

tkhurana commented:
@virajjasani can you also add some unit tests in `WhereOptimizerTest` . Those tests verify that the scan range is generated correctly. I want to make sure that we have some tests for descending columns with like expressions.

jinggou commented:
LGTM. The test case that failed before passes now (e.g., descending columns with LIKE x%).

virajjasani commented:
Thank you for the reviews @stoty @tkhurana @jinggou

Tanuj, sure let me add some test in `WhereOptimizerTest`, sounds good

virajjasani commented:
done

virajjasani commented on phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:
```diff
@@ -1691,8 +1691,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
+            // TODO: is there a case where we'd need to go through the childPart to calculate the key range?
+            PColumn column = childSlot.getKeyPart().getColumn();
+            PDataType type = column.getDataType();
+            byte[] key = PVarchar.INSTANCE.toBytes(startsWith, SortOrder.ASC);
```
interesting! do we support `col1 LIKE ('xy%' || col2)` case? 
i tried this and saw that we return right here due to empty childParts:
```
        @Override
        public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
            // TODO: optimize ILIKE by creating two ranges for the literal prefix: one with lower case, one with upper case
            if (childParts.isEmpty()) {
                return null;
            }
...
...
```

virajjasani replied:
Even `col1 LIKE col2` doesn't seem to be working as childParts are coming empty, resulting in null KeySlots

stoty replied:
Makes sense, we cannot use values from the table for constructing scan filters on the rowkey (unless they are coming from uncorrelated subqueries, but those are effectively constants)

dbwong replied:
Thanks for the test, @virajjasani I do think that in theory `'xy%' || col2 `would be extractable as the prefix here is a constant though I was guessing we don't try to optimize this currently.

virajjasani replied:
Thanks for the review @dbwong
